File: ProblemSet/912.sort-an-array/python/912.sort-an-array 快速排序 单向扫描.py
import random

#
# @lc app=leetcode id=912 lang=python3
#
# [912] Sort an Array
#

# @lc code=start
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        self.quickSort(nums, 0, len(nums) - 1)
        return nums


    def quickSort(self, nums: List[int], left: int, right: int) -> List[int]:
        if left < right:
            pos = self.partition(nums, left, right)
            self.quickSort(nums, left, pos - 1)
            self.quickSort(nums, pos + 1, right)

    
    def partition(self, nums: List[int], left: int, right: int) -> int:
        pivot = random.randint(0, right - left) + left
        nums[left], nums[pivot] = nums[pivot], nums[left]
        i = left
        for j in range(left + 1, right + 1):
            if nums[j] < nums[left]:
                i += 1
                nums[i], nums[j] = nums[j], nums[i]
        nums[left], nums[i] = nums[i], nums[left]
        return i


# @lc code=end

File: ProblemSet/912.sort-an-array/java/912.sort-an-array 快速排序 单向扫描.java
import java.util.Random;

/*
 * @lc app=leetcode id=912 lang=java
 *
 * [912] Sort an Array
 */

// @lc code=start
class Solution {
    public int[] sortArray(int[] nums) {
        quickSort(nums, 0, nums.length - 1);
        return nums;
    }

    private void quickSort(int[] nums, int left, int right) {
        if (left < right) {
            int pos = partition(nums, left, right);
            quickSort(nums, left, pos - 1);
            quickSort(nums, pos + 1, right);
        }
    } 

    private int partition(int[] nums, int left, int right) {
        int pivot = new Random().nextInt(right - left + 1) + left;
        swap(nums, left, pivot);
        int i = left;
        for (int j = left + 1; j <= right; ++j) {
            if (nums[j] < nums[left]) 
                swap(nums, ++i, j);
        }
        swap(nums, left, i);
        return i;
    }

    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

}
// @lc code=end

